presto-hive/src/main/java/com/facebook/presto/twitter/hive/thrift/ThriftGenericRow.java

```diff
@@ -0,0 +1,305 @@
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.protocol.TField;
+import org.apache.thrift.protocol.TList;
+import org.apache.thrift.protocol.TMap;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.protocol.TProtocolUtil;
+import org.apache.thrift.protocol.TSet;
+import org.apache.thrift.protocol.TStruct;
+import org.apache.thrift.protocol.TType;
+import org.apache.thrift.transport.TMemoryInputTransport;
+import org.apache.thrift.transport.TTransport;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ThriftGenericRow implements TBase<ThriftGenericRow, ThriftGenericRow.Fields>
+{
+    private static final Logger log = Logger.get(ThriftGenericRow.class);
+    private final Map<Short, Object> values = new HashMap<>();
+    private byte[] buf = null;
+    private int off = 0;
+    private int len = 0;
+
+    public ThriftGenericRow() {}
+
+    public ThriftGenericRow(Map<Short, Object> values)
+    {
+        this.values.putAll(values);
+    }
+
+    public class Fields implements TFieldIdEnum
+    {
+        private final short thriftId;
+        private final String fieldName;
+
+        Fields(short thriftId, String fieldName)
+        {
+            this.thriftId = thriftId;
+            this.fieldName = fieldName;
+        }
+
+        public short getThriftFieldId()
+        {
+            return thriftId;
+        }
+
+        public String getFieldName()
+        {
+            return fieldName;
+        }
+    }
+
+    public void read(TProtocol iprot) throws TException
+    {
+        TTransport trans = iprot.getTransport();
+        buf = trans.getBuffer();
+        off = trans.getBufferPosition();
+        TProtocolUtil.skip(iprot, TType.STRUCT);
+        len = trans.getBufferPosition() - off;
+    }
+
+    public void parse() throws TException
+    {
+        parse(null);
+    }
+
+    public void parse(short[] thriftIds) throws TException
+    {
+        Set<Short> idSet = thriftIds == null ? null : new HashSet(Arrays.asList(ArrayUtils.toObject(thriftIds)));
+        TMemoryInputTransport trans = new TMemoryInputTransport(buf, off, len);
+        TBinaryProtocol iprot = new TBinaryProtocol(trans);
+        TField field;
+        iprot.readStructBegin();
+        while (true) {
+            field = iprot.readFieldBegin();
+            if (field.type == TType.STOP) {
+                break;
+            }
+            if (idSet != null && !idSet.remove(Short.valueOf(field.id))) {
+                TProtocolUtil.skip(iprot, field.type);
+            }
+            else {
+                values.put(field.id, readElem(iprot, field.type));
+            }
+            iprot.readFieldEnd();
+        }
+        iprot.readStructEnd();
+    }
+
+    private Object readElem(TProtocol iprot, byte type) throws TException
```

Comment:
Can you create a unit test for this class/method that constructs a thrift object of all supported (and unsupported) types and then asserts that it can read (or fail appropriately) for each.